### PR TITLE
Bootstrap the Consul ACL system if no token is given

### PR DESCRIPTION
**Pull request by remilapeyre**
Being able to bootstrap the ACL system of Consul is something that has
long been asked of its Terraform provider (https://github.com/hashicorp/terraform-provider-consul/issues/95).

We always refused to implement a solution that would save this token in
the Terraform state has the new ACL system in 1.4 meant that we could
finally referenced some token without having access to their secret ID.
Storing the bootstrap token in the state would have made this useless
and would potentially be a security issue.

This change makes it possible to configure a new Consul secret engine
without providing a token, in that case Vault knows that the ACL system
has not yet been boostraped and do it itself. This means that will at
last be able to have completely automatic and secure Consul cluster
creation using Terraform, this has been wanted by our users for some
time now.

### CODE REVIEW

**Comment by mike-sol** [comment]
Is there any hope for this to get merged? Would really love this functionality to be able to programmatically and idempotently bootstrap consul ACLs directly from a Terraform / Vault setup.

**Comment by remilapeyre** [comment]
Hi @robmonte, conflicts should be fixed now :)

File: helper/testhelpers/consul/consulhelper.go [hunk 1]
```diff
@@ -113,101 +118,105 @@ func PrepareTestContainer(t *testing.T, version string, isEnterprise bool) (func
...
-			ACLList := []*consulapi.ACLLink{{Name: "test"}}
-
-			role := &consulapi.ACLRole{
-				Name:        "role-test",
-				Description: "consul roles test",
-				Policies:    ACLList,
+				service_prefix "" {
+					policy = "read"
+				}
+		`,
```
Comment (robmonte) [inline helper/testhelpers/consul/consulhelper.go:138]: Can we close this quote back on line 137?

File: website/content/docs/secrets/consul.mdx [hunk 2]
```diff
@@ -25,41 +25,52 @@ management tool.
     By default, the secrets engine will mount at the name of the engine. To
     enable the secrets engine at a different path, use the `-path` argument.
 
-1.  Bootstrap the Consul ACL system if not already done. To begin configuring the secrets engine, we must give Vault
-    the necessary credentials to manage Consul.
+1. Vault can bootstrap the ACL system of your Consul cluster if it has
```
Comment (robmonte) [inline website/content/docs/secrets/consul.mdx:28]: Appreciate the docs update included! I think that due to the way this new feature will create the bootstrap token _and_ store it in Vault without exposing it to the user, we'll need to elaborate a bit more about this process.

I made a task for myself to add more documentation at a later date for this, but I'm happy with this current version for now!

**Review by robmonte — CHANGES_REQUESTED** [review]
Just a couple small change requests and some additional comments. Looks great!!

File: builtin/logical/consul/backend_test.go [hunk 3]
```diff
@@ -22,24 +22,32 @@ func TestBackend_Config_Access(t *testing.T) {
...
+			testBackendConfigAccess(t, "1.3.1", true)
 		})
 		t.Run("post-1.4.0", func(t *testing.T) {
 			t.Parallel()
-			testBackendConfigAccess(t, "")
+			testBackendConfigAccess(t, "", true)
+		})
+		t.Run("pre-1.4.0 automatic-bootstrap", func(t *testing.T) {
+			t.Parallel()
+			testBackendConfigAccess(t, "1.3.0", false)
```
Comment (robmonte) [inline builtin/logical/consul/backend_test.go:33]: Can this line be updated to use version 1.3.1 to match the above test case?

**Comment by robmonte** [comment]
@remilapeyre I left a PR review with some feedback and small change requests. If you have any responses to them, please let me know.

To avoid potential upcoming merge conflicts I am planning on getting this merged in very soon. If you don't have time to go through my review, I can move your commits into a new PR and make the updates myself on top of your work so you retain credit for your contribution. I'll wait a couple days before doing so.

File: builtin/logical/consul/path_config.go [hunk 4]
```diff
@@ -123,3 +140,15 @@ type accessConfig struct {
...
+	consulConf := api.DefaultNonPooledConfig()
+	consulConf.Address = conf.Address
+	consulConf.Scheme = conf.Scheme
+	consulConf.Token = conf.Token
+	consulConf.TLSConfig.CAPem = []byte(conf.CACert)
+	consulConf.TLSConfig.CertPEM = []byte(conf.ClientCert)
+	consulConf.TLSConfig.KeyPEM = []byte(conf.ClientKey)
+
+	return api.NewClient(consulConf)
+}
```
[the remark this answers is not in the document]
Comment (remilapeyre) [reply builtin/logical/consul/path_config.go]: Thanks, I updated this function :)

File: helper/testhelpers/consul/consulhelper.go [hunk 5]
```diff
@@ -94,6 +94,11 @@ func PrepareTestContainer(t *testing.T, version string, isEnterprise bool) (func
 			return nil, err
 		}
 
+		// Make sure Consul is up
+		if _, err = consul.Status().Leader(); err != nil {
```
[the remark this answers is not in the document]
Comment (remilapeyre) [reply helper/testhelpers/consul/consulhelper.go:98]: Yes, when the tests run before the leader has been elected they fail on my computer, it's common enough that I needed to add this line.

**Comment by remilapeyre** [comment]
Hi @robmonte, I updated the PR based on your comments. Everything should be ready now.

**Review by robmonte — APPROVED** [review]

**Pull request merged**